js/src/vector/arrow.ts

```diff
@@ -0,0 +1,245 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import * as Schema_ from '../format/Schema_generated';
+import * as Message_ from '../format/Message_generated';
+import Field = Schema_.org.apache.arrow.flatbuf.Field;
+import FieldNode = Message_.org.apache.arrow.flatbuf.FieldNode;
+
+import { Vector } from './vector';
+import { Utf8Vector as Utf8VectorBase } from './utf8';
+import { StructVector as StructVectorBase } from './struct';
+import { DictionaryVector as DictionaryVectorBase } from './dictionary';
+import {
+    ListVector as ListVectorBase,
+    BinaryVector as BinaryVectorBase,
+    FixedSizeListVector as FixedSizeListVectorBase
+} from './list';
+
+import {
+    BoolVector as BoolVectorBase,
+    Int8Vector as Int8VectorBase,
+    Int16Vector as Int16VectorBase,
+    Int32Vector as Int32VectorBase,
+    Int64Vector as Int64VectorBase,
+    Uint8Vector as Uint8VectorBase,
+    Uint16Vector as Uint16VectorBase,
+    Uint32Vector as Uint32VectorBase,
+    Uint64Vector as Uint64VectorBase,
+    Float16Vector as Float16VectorBase,
+    Float32Vector as Float32VectorBase,
+    Float64Vector as Float64VectorBase,
+    Date32Vector as Date32VectorBase,
+    Date64Vector as Date64VectorBase,
+    Time32Vector as Time32VectorBase,
+    Time64Vector as Time64VectorBase,
+    DecimalVector as DecimalVectorBase,
+    TimestampVector as TimestampVectorBase,
+} from './numeric';
+
+import { nullableMixin, fieldMixin } from './traits';
+
+function MixinArrowTraits<T extends Vector<any>, TArgv>(
+    Base: new (argv: TArgv) => T,
+    Field: new (argv: TArgv & { field: Field, fieldNode: FieldNode }) => T,
+    Nullable: new (argv: TArgv & { validity: Uint8Array }) => T,
+    NullableField: new (argv: TArgv & { validity: Uint8Array, field: Field, fieldNode: FieldNode }) => T,
+) {
```

Comment:
@TheNeuralBit yeah so the ES6 class spec states that the `name` property of a class constructor is immutable (and they're also not allowed to be computed properties, like `let x = 'myClass'; class [x] extends Foo {}`). And anonymous class names default to "Object", reading as `Object { data: Int32Array }` instead of `Int32Vector` when debugging. While this is ugly and hard to scale if we want to add more mixin behaviors, I figure it's a win for anyone using the library in the real world to see descriptive class names.

Reply:
Ah makes sense, I figured there must be a good reason. This seems like a great application for some kind of preprocessor with macros or a code generator... but I don't know of any that would integrate well with JS build/dev tools

Reply:
@TheNeuralBit yeah we could use [babel-codegen](https://github.com/kentcdodds/babel-plugin-codegen), [babel-preval](https://github.com/kentcdodds/babel-plugin-preval), or [babel-macros](https://github.com/kentcdodds/babel-macros) if we want. I was hoping to avoid babel if possible, but since we're webpacking the es2015+ UMD bundles anyway, it wouldn't be too much of a headache.

Reply:
@TheNeuralBit but if we do want to do more compilation steps beyond what the TS compiler does, it'd be neat to also run [preval on the flatbuffers generated code](https://gist.github.com/trxcllnt/84bb4893b6db957925ed7625fd0f34e5)